src/network/operator.rs

```diff
@@ -0,0 +1,335 @@
+                        if is_template_stale {
+                            // Construct a new block template.
+                            let transactions = operator.memory_pool.read().await.transactions();
+                            let (block_template, _) = operator
+                                .ledger_reader
+                                .get_block_template(recipient, E::COINBASE_IS_PUBLIC, &transactions, &mut thread_rng())
+                                .expect("Should be able to generate a block template");
+
+                            // Acquire the write lock to update the block template.
+                            *operator.block_template.write().await = Some(block_template);
+                        }
+
+                        // Sleep for `5` seconds.
+                        tokio::time::sleep(Duration::from_secs(5)).await;
+                    }
+                }));
+                // Wait until the operator handler is ready.
+                let _ = handler.await;
+            } else {
+                error!("Missing operator address. Please specify an Aleo address in order to operate a pool");
+            }
+        }
+
+        Ok(operator)
+    }
+
+    /// Returns an instance of the operator router.
+    pub fn router(&self) -> OperatorRouter<N> {
+        self.operator_router.clone()
+    }
+
+    /// Returns all the shares in storage.
+    pub fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.state.to_shares()
+    }
+
+    ///
+    /// Performs the given `request` to the operator.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: OperatorRequest<N>) {
+        match request {
+            OperatorRequest::PoolRegister(peer_ip, address) => {
+                if let Some(block_template) = &*self.block_template.read().await {
+                    // Ensure this prover exists in the list first, and retrieve their share difficulty.
+                    let share_difficulty = self
+                        .provers
+                        .write()
+                        .await
+                        .entry(address)
+                        .or_insert((Instant::now(), BASE_SHARE_DIFFICULTY, 0))
+                        .1;
+
+                    // Route a `PoolRequest` to the peer.
+                    let message = Message::PoolRequest(share_difficulty, Data::Object(block_template.clone()));
+                    if let Err(error) = self.peers_router.send(PeersRequest::MessageSend(peer_ip, message)).await {
+                        warn!("[PoolRequest] {}", error);
+                    }
+                } else {
+                    warn!("[PoolRegister] No current block template exists");
+                }
+            }
+            OperatorRequest::PoolResponse(peer_ip, block, prover_address) => {
+                if let Some(block_template) = &*self.block_template.read().await {
```

Comment:
```suggestion
                if let Some(block_template) = self.block_template.read().await.cloned() {
```
same reasoning as earlier